File: backend/app/account_structure.py
"""
Taxonomía del plan de cuentas (categoría matriz → tipo → detalle).

Fuente de verdad compartida con ``frontend/src/features/accounting/accountStructure.js``.
Los valores ``account_type`` persistidos en BD son los tipos ledger en inglés (QuickBooks).
"""

from __future__ import annotations

from typing import TypedDict


class AccountGroup(TypedDict, total=False):
    tipo: str
    account_type: str
    detalles: list[str]
    uses_payment_methods: bool


EFECTIVO_EQUIVALENTES_TIPO = "Efectivo y equivalentes"

# Categoría matriz → grupos (tipo de cuenta → detalles permitidos).
ACCOUNT_STRUCTURE: dict[str, list[AccountGroup]] = {
    "ACTIVOS": [
        {
            "tipo": "Activos Corrientes",
            "account_type": "asset",
            "detalles": [
                "Inventario",
                "Fondos sin depositar",
                "Anticipo empleados",
            ],
        },
        {
            "tipo": EFECTIVO_EQUIVALENTES_TIPO,
            "account_type": "asset",
            "detalles": [],
            "uses_payment_methods": True,
        },
    ],
    "RESPONSABILIDAD": [
        {
            "tipo": "Cuentas por pagar",
            "account_type": "liability",
            "detalles": [
                "Cuentas por pagar",
                "Anticipos de clientes",
                "Saldos a favor",
            ],
        },
    ],
    "INGRESOS": [
        {
            "tipo": "Ingresos",
            "account_type": "income",
            "detalles": [
                "Venta de productos y Servicios",
                "Ingresos recarga de saldo",
                "Otros ingresos principales",
            ],
        },
    ],
    "GASTO": [
        {
            "tipo": "Gastos",
            "account_type": "expense",
            "detalles": [
                "Deudas incobrables",
                "Gastos administrativos",
                "Gasto nómina",
                "Tasas y comisiones",
                "Publicidad y Promoción",
                "Reparación y mantenimiento",
                "Suministros y materiales",
                "Comida y ocio",
                "Servicios varios",
            ],
        },
        {
            "tipo": "Costos de venta",
            "account_type": "cost_of_sales",
            "detalles": ["Descuentos", "Otros"],
        },
        {
            "tipo": "Otros gastos",
            "account_type": "expense",
            "detalles": [
                "Pérdida de cambio",
                "Otros gastos",
                "Liquidaciones",
            ],
        },
    ],
}

# Alias legacy (BD antigua / migraciones) → detalle canónico actual.
LEGACY_DETAIL_TYPE_ALIASES: dict[str, str] = {
    "Gasto nomina": "Gasto nómina",
    "Publicidad y Promocion": "Publicidad y Promoción",
    "Reparacion y mantenimiento": "Reparación y mantenimiento",
    "Comida y osio": "Comida y ocio",
    "Perdida de cambio": "Pérdida de cambio",
    "otros gastos": "Otros gastos",
    "otros": "Otros",
    "descuentos": "Descuentos",
    "Ganancia por tipo de cambio": "Otros ingresos principales",
}

LIQUID_DEPOSIT_DETAIL_TYPES: frozenset[str] = frozenset({"Fondos sin depositar"})


def group_uses_payment_methods(tipo: str) -> bool:
    for grupos in ACCOUNT_STRUCTURE.values():
        for g in grupos:
            if g.get("tipo") == tipo:
                return bool(g.get("uses_payment_methods"))
    return False


def is_efectivo_equivalentes_detail(*, linked_payment_method: str | None) -> bool:
    return bool((linked_payment_method or "").strip())


def normalize_detail_type(value: str | None) -> str | None:
    if value is None:
        return None
    s = str(value).strip()
    if not s:
        return None
    return LEGACY_DETAIL_TYPE_ALIASES.get(s, s)


def all_detail_types() -> frozenset[str]:
    out: set[str] = set()
    for grupos in ACCOUNT_STRUCTURE.values():
        for g in grupos:
            if g.get("uses_payment_methods"):
                continue
            out.update(g.get("detalles") or [])
    return frozenset(out)


def detail_types_for_ledger(account_type: str) -> frozenset[str]:
    out: set[str] = set()
    for grupos in ACCOUNT_STRUCTURE.values():
        for g in grupos:
            if g.get("uses_payment_methods"):
                continue
            if g.get("account_type") == account_type:
                out.update(g.get("detalles") or [])
    return frozenset(out)


def validate_linked_payment_method_name(db, name: str | None) -> str:
    """Comprueba que el nombre exista en ``payment_methods`` activos. Devuelve nombre normalizado."""
    from app.models.payment_method import PaymentMethod

    raw = (name or "").strip()
    if not raw:
        raise ValueError("Selecciona un método de pago para la cuenta de efectivo y equivalentes.")
    row = (
        db.query(PaymentMethod)
        .filter(
            PaymentMethod.is_active.is_(True),
            PaymentMethod.name == raw,
        )
        .first()
    )
    if row is None:
        raise ValueError(f"El método de pago «{raw}» no existe o está inactivo.")
    return str(row.name)


def validate_chart_account_classification(
    *,
    account_type: str,
    detail_type: str | None,
    linked_payment_method: str | None = None,
) -> None:
    """Valida par (account_type, detail_type) contra la taxonomía maestra. Lanza ValueError."""
    dt_raw = (detail_type or "").strip()
    lp = (linked_payment_method or "").strip()

    if lp:
        if account_type != "asset":
            raise ValueError("Solo las cuentas de activo pueden vincular un método de pago.")
        # Efectivo y equivalentes: detalle dinámico (= nombre del método de pago).
        return

    if not dt_raw:
        raise ValueError("Indica el tipo de detalle de la cuenta.")

    dt = normalize_detail_type(dt_raw) or dt_raw
    allowed = detail_types_for_ledger(account_type)
    if dt not in allowed:
        allowed_txt = ", ".join(sorted(allowed))
        raise ValueError(
            f"Tipo de detalle «{dt_raw}» no permitido para account_type «{account_type}». "
            f"Valores válidos: {allowed_txt}."
        )

File: frontend/src/features/accounting/accountStructure.js
/**
 * Taxonomía del plan de cuentas.
 * Mantener sincronizado con backend/app/account_structure.py
 */
export const EFECTIVO_EQUIVALENTES_TIPO = 'Efectivo y equivalentes'

export const ACCOUNT_STRUCTURE = {
  ACTIVOS: [
    {
      tipo: 'Activos Corrientes',
      accountType: 'asset',
      detalles: ['Inventario', 'Fondos sin depositar', 'Anticipo empleados'],
    },
    {
      tipo: EFECTIVO_EQUIVALENTES_TIPO,
      accountType: 'asset',
      detalles: [],
      usesPaymentMethods: true,
    },
  ],
  RESPONSABILIDAD: [
    {
      tipo: 'Cuentas por pagar',
      accountType: 'liability',
      detalles: ['Cuentas por pagar', 'Anticipos de clientes', 'Saldos a favor'],
    },
  ],
  INGRESOS: [
    {
      tipo: 'Ingresos',
      accountType: 'income',
      detalles: [
        'Venta de productos y Servicios',
        'Ingresos recarga de saldo',
        'Otros ingresos principales',
      ],
    },
  ],
  GASTO: [
    {
      tipo: 'Gastos',
      accountType: 'expense',
      detalles: [
        'Deudas incobrables',
        'Gastos administrativos',
        'Gasto nómina',
        'Tasas y comisiones',
        'Publicidad y Promoción',
        'Reparación y mantenimiento',
        'Suministros y materiales',
        'Comida y ocio',
        'Servicios varios',
      ],
    },
    {
      tipo: 'Costos de venta',
      accountType: 'cost_of_sales',
      detalles: ['Descuentos', 'Otros'],
    },
    {
      tipo: 'Otros gastos',
      accountType: 'expense',
      detalles: ['Pérdida de cambio', 'Otros gastos', 'Liquidaciones'],
    },
  ],
}

export const LEGACY_DETAIL_TYPE_ALIASES = {
  'Gasto nomina': 'Gasto nómina',
  'Publicidad y Promocion': 'Publicidad y Promoción',
  'Reparacion y mantenimiento': 'Reparación y mantenimiento',
  'Comida y osio': 'Comida y ocio',
  'Perdida de cambio': 'Pérdida de cambio',
  'otros gastos': 'Otros gastos',
  otros: 'Otros',
  descuentos: 'Descuentos',
  'Ganancia por tipo de cambio': 'Otros ingresos principales',
}

export const TIPO_CUENTA_SEP = ':::'

export function normalizeDetailType(value) {
  if (value == null) return ''
  const s = String(value).trim()
  if (!s) return ''
  return LEGACY_DETAIL_TYPE_ALIASES[s] ?? s
}

export function buildTipoCuentaValue(categoriaMatriz, tipoEtiqueta) {
  return `${categoriaMatriz}${TIPO_CUENTA_SEP}${tipoEtiqueta}`
}

export function parseTipoCuentaValue(value) {
  if (!value || typeof value !== 'string') return { categoriaMatriz: '', tipo: '' }
  const i = value.indexOf(TIPO_CUENTA_SEP)
  if (i === -1) return { categoriaMatriz: '', tipo: '' }
  return {
    categoriaMatriz: value.slice(0, i),
    tipo: value.slice(i + TIPO_CUENTA_SEP.length),
  }
}

export function findGrupoByTipoCuentaValue(composite) {
  const { categoriaMatriz, tipo } = parseTipoCuentaValue(composite)
  const list = ACCOUNT_STRUCTURE[categoriaMatriz]
  if (!Array.isArray(list)) return null
  return list.find((g) => g.tipo === tipo) ?? null
}

export function usesPaymentMethodsForTipoCuenta(composite) {
  const g = findGrupoByTipoCuentaValue(composite)
  return Boolean(g?.usesPaymentMethods)
}

export function isEfectivoYEquivalentesTipoCuenta(composite) {
  const { tipo } = parseTipoCuentaValue(composite)
  return tipo === EFECTIVO_EQUIVALENTES_TIPO
}

/** Detalles estáticos del grupo; vacío si el grupo usa métodos de pago dinámicos. */
export function getDetallesForTipoCuenta(composite) {
  const g = findGrupoByTipoCuentaValue(composite)
  if (g?.usesPaymentMethods) return []
  return Array.isArray(g?.detalles) ? [...g.detalles] : []
}

export function getBackendAccountTypeForTipoCuenta(composite) {
  const g = findGrupoByTipoCuentaValue(composite)
  return g?.accountType ?? 'income'
}

export function getDefaultTipoCuentaValue() {
  const firstCat = Object.keys(ACCOUNT_STRUCTURE)[0]
  const firstGrupo = ACCOUNT_STRUCTURE[firstCat]?.[0]
  if (!firstCat || !firstGrupo?.tipo) return ''
  return buildTipoCuentaValue(firstCat, firstGrupo.tipo)
}

export function getFirstDetalleForTipoCuenta(composite, paymentMethodNames = []) {
  if (usesPaymentMethodsForTipoCuenta(composite)) {
    return paymentMethodNames[0] || ''
  }
  const d = getDetallesForTipoCuenta(composite)
  return d.length ? d[0] : ''
}

export function inferTipoCuentaFromApi(accountType, detailType, linkedPaymentMethod = null) {
  const lp = linkedPaymentMethod != null ? String(linkedPaymentMethod).trim() : ''
  if (accountType === 'asset' && lp) {
    return buildTipoCuentaValue('ACTIVOS', EFECTIVO_EQUIVALENTES_TIPO)
  }

  const dtNorm = normalizeDetailType(detailType)
  let fallback = null
  for (const [categoriaMatriz, grupos] of Object.entries(ACCOUNT_STRUCTURE)) {
    for (const g of grupos) {
      if (g.accountType !== accountType) continue
      if (g.usesPaymentMethods) continue
      if (fallback === null) {
        fallback = buildTipoCuentaValue(categoriaMatriz, g.tipo)
      }
      if (dtNorm && Array.isArray(g.detalles)) {
        const match = g.detalles.find((d) => normalizeDetailType(d) === dtNorm || d === detailType)
        if (match) {
          return buildTipoCuentaValue(categoriaMatriz, g.tipo)
        }
      }
    }
  }
  return fallback ?? getDefaultTipoCuentaValue()
}

const LIQUID_DEPOSIT_DETAIL_TYPES = new Set(['Fondos sin depositar'])

export function isLiquidDepositChartAccount(account) {
  if (!account || account.account_type !== 'asset') return false
  const dt = normalizeDetailType(account.detail_type)
  if (LIQUID_DEPOSIT_DETAIL_TYPES.has(dt)) return true
  const lp = String(account.linked_payment_method ?? '').trim()
  return lp.length > 0
}

export function isInventoryChartAccount(account) {
  if (!account || account.account_type !== 'asset') return false
  const dt = normalizeDetailType(account.detail_type)
  return dt === 'Inventario' || String(account.detail_type ?? '').trim().toLowerCase() === 'inventario'
}

export function sortPaymentMethodNames(list) {
  return [...(Array.isArray(list) ? list : [])]
    .filter((m) => m?.is_active !== false)
    .map((m) => String(m?.name || '').trim())
    .filter(Boolean)
    .sort((a, b) => a.localeCompare(b, 'es', { sensitivity: 'base' }))
}
